core/kernel/console.c

```diff
@@ -129,7 +129,7 @@ void configure_console_from_dt(void)
 		return;
 
 	dt_drv = dt_find_compatible_driver(fdt, offs);
-	if (!dt_drv)
+	if (!dt_drv || dt_drv->type != DT_DRIVER_UART)
```

Comment:
Agreed, this *should* not return a driver other than uart but if the phandle passed as stdout-path is not pointing to a UART driver (but a clock one for instance) this will crash badly since  there is a cast after that:
```
	sdrv = (const struct serial_driver *)dt_drv->driver;
```
But the driver field depends on the driver type. We could also add more ELF section for each type of driver such as the kernel does.

Reply:
Indeed, for clocks, the check is needed since I want to iterate on the clocks only. As said, another solution would be to have a per driver type sections such as Linux does. It would probably be cleaner but would lead to multiple sections.
Regarding the check for DT_DRIVER_UART, if the devicetree is correct, then yes, there is no need for it, but if the DT is wrong and reference a wrong node instead of a serial then it could be problematic. I'll do as you prefer !

Reply:
Ok, I wanted to make sure this solution was accepted by maintainers.